### ISSUE TITLE
`GDScript` resources are not unloaded due to an extra reference

### ISSUE THREAD

**Issue by dalexeev** [godotengine/godot#77513]
### Godot version

v4.0.3.stable.official [5222a99f5], master

### System information

Kubuntu 23.04

### Issue description

It looks like we have an extra reference to the script somewhere, which prevents it from being freed at the right time. 

See also [the comment](https://github.com/godotengine/godot/pull/77129#discussion_r1205894765).

CC @adamscott

### Steps to reproduce

```gdscript
extends Node

const RES_PATH = "res://a.gd"
#const RES_PATH = "res://icon.svg"

var i := 0
var wr: WeakRef

func _on_timer_timeout() -> void:
    print("===")
    if i % 2:
        print("wr = ", wr.get_ref())
    else:
        var res = load(RES_PATH)
        wr = weakref(res)
        print("ref count = ", res.get_reference_count())
        print("wr = ", wr.get_ref())
    i += 1
```

```
===
ref count = 2
wr = <GDScript#-9223372008484502375>
===
wr = <GDScript#-9223372008484502375>
===
ref count = 2
wr = <GDScript#-9223372008484502375>
===
wr = <GDScript#-9223372008484502375>
```

```
===
ref count = 1
wr = <CompressedTexture2D#-9223372010162223986>
===
wr = <null>
===
ref count = 1
wr = <CompressedTexture2D#-9223372009843456882>
===
wr = <null>
```

### Minimal reproduction project

[bug-script-ref-count.zip](https://github.com/godotengine/godot/files/11575495/bug-script-ref-count.zip)

**Comment by adamscott** [comment]
> It looks like we have an extra reference to the script somewhere, which prevents it from being freed at the right time.

It's by design to enable cyclic references. Each GDScript loaded has a Ref instance living here:
https://github.com/godotengine/godot/blob/e188d619227990001667821dac8bc8940076d4a9/modules/gdscript/gdscript_cache.h#L79-L81

But the garbage collection part is not yet implemented, unfortunately.

**Comment by dalexeev** [comment]
Does this mean that `@unload_static` doesn't really matter now: scripts aren't unloaded anyway?

**Comment by vnen** [comment]
> It's by design to enable cyclic references. Each GDScript loaded has a Ref instance living [in the cache]

I believe we talked about this in the chat but let me register here as well for the record. The cache should not be keeping extra references to scripts. It's fine if the scripts themselves are in a cycle, since that is unavoidable (and where a cycle collector would be handy), but in the case there are no cycles the scripts should be naturally freed.

Scripts sometimes preloads scenes which can get quite heavy so keeping them in memory can ended up with a lot of used space which is not obvious where it comes from. Without the collector it's much worse because there's currently no way to unload scripts.

So this needs to be addressed. I do plan to work on this (probably next month if nobody beats me to it). I've also notice some other issues with the cache, like keeping non-compiled scripts as if they were compiled which gives another sort of problems.

**Comment by adamscott** [comment]
> The cache should not be keeping extra references to scripts.

I agree. I was the one that modified the cache to make it so that it keeps an extra reference to the script.

But the reason why references are kept in the cache is because a script reference count would drop to zero even before having the opportunity to cyclically reference itself. So if we're able to fix that issue, I'm all in to drop these references.

Maybe we could try to store the refs shallow scripts, but once a script is fully loaded, only cache the pointer (not a ref of a script).

**Comment by jinyangcruise** [comment]
Looking forward to any progress on this👀

**Comment by MikeSchulze** [comment]
After I convert an internal class that holds variables to static members, I run now into this issue.
Adding the suggested `@static_unload` is not working

This bug makes using static members unusable!
I'm just wondering why this critical bug is still ignoring....

Also try to add a workaround do not help.
```gd
func _notification(what: int) -> void:
	if what == NOTIFICATION_PREDELETE:
		prints("delete static memebers")
                ...
```


This Bug still occurs in v4.5.stable.mono.official [876b29033]
Please fix this Bug as soon is possible, this makes using static variables unusable.

**Comment by dalexeev** [comment]
> I'm just wondering why this critical bug is still ignoring....

@MikeSchulze I understand your frustration, but we really don't know how to fix this without hurting GDScript compilation performance.

The problem is that GDScript resources have a complex and non-linear loading order. Scripts aren't compiled all at once, but are loaded, parsed, analyzed, and compiled on demand, as a `GDScript` resource is loaded by `ResourceLoader` (full scripts) or required as a dependency by other scripts (shallow scripts). `GDScriptCache` stores references to all full and shallow scripts to avoid re-parsing and re-analyzing the same dependencies each time a next script is loaded. And since script loading can occur dynamically over a long period of time, `GDScriptCache` simply can't know when it's safe to stop storing references to scripts.

Also, in my opinion, the caching/loading system itself is rather convoluted and somewhat overcomplicated. But even if we fix some of the issues with the current cache system, the "script is a resource" approach and the dynamic script loading order in Godot make it unlikely that we'll be able to avoid keeping a script reference in the cache without degrading GDScript compilation performance.

There's currently ongoing discussion about potential major changes to GDScript. Perhaps in the future, we'll decide to abandon the "script is a resource" approach. This option would fix this and other issues.

**Comment by MikeSchulze** [comment]
@dalexeev,
Thank you for your detailed response. It's good to shed more light on this topic.
If I understand you correctly, using statics is currently not safe.
I would say it would be helpful to provide users with a tool that allows them to release this manually, similar to the “free” command, but for scripts.
For example, “dispose.”
I am currently forced to roll back all my changes to statics and use my previous workaround.

**Comment by dalexeev** [comment]
You can reset your static variables in a static method (assign `null` to `RefCounted`, call `queue_free()` or `free()` on nodes and objects, etc.). The script itself won't be deleted, but you can still clear the static data.

**Comment by MikeSchulze** [comment]
> You can reset your static variables in a static method (assign `null` to `RefCounted`, call `queue_free()` or `free()` on nodes and objects, etc.). The script itself won't be deleted, but you can still clear the static data.

Nope, that not work!
```

func _notification(what: int) -> void:
	if what == NOTIFICATION_PREDELETE:
		__dispose()

static func __dispose() -> void:
	__excluded_methods__.clear()
	__verifier_instance__ = null
	__return_values__.clear()
	__return_value__ = null
```
When I use my workaround and put the static variables in an inner class and use the engine metadata storage instead of static, it works fine.
```
func __init(__script: GDScript, mock_working_mode: String) -> void:
	Engine.set_meta(__INSTANCE_ID, GdUnitMockState.new(mock_working_mode))


static func __mock_state() -> GdUnitMockState:
	return Engine.get_meta(__INSTANCE_ID)


func _notification(what: int) -> void:
	if what == NOTIFICATION_PREDELETE and Engine.has_meta(__INSTANCE_ID):
		Engine.remove_meta(__INSTANCE_ID)
```

**Comment by dalexeev** [comment]
I don't know the details of your issue, but as far as I know, static variables are safe and are no different in this regard from non-static variables, metadata, arrays, etc. If you're encountering memory leaks with static variables, it's most likely due to a memory management error in your code (leftover references to `RefCounted` data, lost orphan nodes, etc.).

If you think there's a bug with static variables, it deserves a separate issue, as this thread is about an "extra" reference to the script itself from `GDScriptCache`.

**Comment by dalexeev** [comment]
> Also, in my opinion, the caching/loading system itself is rather convoluted and somewhat overcomplicated. But even if we fix some of the issues with the current cache system, the "script is a resource" approach and the dynamic script loading order in Godot make it unlikely that we'll be able to avoid keeping a script reference in the cache without degrading GDScript compilation performance.

I'd like to clarify that this isn't entirely true. It's theoretically possible to maintain the "script is a resource" approach without storing a reference to the script in `GDScriptCache`. It doesn't actually require the `GDScript` resource itself, only the abstract syntax tree, the bytecode of the compiled functions, and other metadata about the class and its members.

But the problem is that the latter two are currently tightly coupled to `GDScript`, `GDScriptFunction`, and `GDScriptParser::DataType`. Separating them is a rather complex technical challenge, likely comparable to the major GDScript update plans being discussed.
